ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,291 @@
+                denominators[i] *= evaluation_points[i] - evaluation_points[j];
+            }
+        }
+        if i + 1 < evaluation_points.len() {
+            for j in i + 1..evaluation_points.len() {
+                denominators[i] *= evaluation_points[i] - evaluation_points[j];
+            }
+        }
+        denominators[i] = denominators[i].invert();
+    }
+}
+
+/// computes the lagrange base polynomials
+/// for correctness, the denominators need to be consistent with the evaluation_points
+/// overwrites `base`
+pub fn compute_lagrange_base<F>(
+    points: &[F],
+    evaluation_points: &[F],
+    denominators: &[F],
+    base: &mut Vec<Vec<F>>,
+) -> ()
+where
+    F: Field,
+{
+    debug_assert_eq!(evaluation_points.len(), denominators.len());
+    #[cfg(debug_assertions)]
+    {
+        for base in base.iter() {
+            debug_assert_eq!(evaluation_points.len(), base.len());
+        }
+    }
+    debug_assert_eq!(points.len(), base.len());
+
+    // for each point, compute base polynomials
+    points
+        .iter()
+        .zip(base.iter_mut())
+        .for_each(|(point, base)| {
+            // set up the denominators
+            base.copy_from_slice(denominators);
+            // compute the nominator
+            for i in 0..evaluation_points.len() {
+                if i > 0 {
+                    for j in 0..i {
+                        base[i] *= *point - evaluation_points[j];
+                    }
+                }
+                if i + 1 < evaluation_points.len() {
+                    for j in i + 1..evaluation_points.len() {
+                        base[i] *= *point - evaluation_points[j];
+                    }
+                }
+            }
+        });
+}
+
+/// precomputed lagrange evaluation
+/// evaluates polynomial using base polynomials
+/// multiplies `result` with input `result`, i.e. use input `F::ONE`
+pub fn lagrange_evaluation_precomputed<F>(y: &[F], base: &Vec<Vec<F>>, result: &mut [F]) -> ()
+where
+    F: Field,
+{
+    debug_assert_eq!(base.len(), result.len());
+    #[cfg(debug_assertions)]
+    {
+        for base in base.iter() {
+            debug_assert_eq!(y.len(), base.len());
+        }
+    }
+    result
+        .iter_mut()
+        .zip(base.iter())
+        .for_each(|(result, base)| {
+            *result *= base
+                .iter()
+                .zip(y.iter())
+                .fold(F::ZERO, |acc, (base, y)| acc + *base * *y)
+        });
+}
+
+/// lagrange evaluation
+/// given points `(x_0,y_0),...,(x_n,y_0)` on a polynomial `p`
+/// and points `p_0,...,p_m`
+/// compute `result_0=result_0*p(p_0),...,result_m=result_m*p(p_m)`
+/// It uses the lagrange method to compute the points `https://en.wikipedia.org/wiki/Lagrange_polynomial`
+/// further, rather than outputting `p(p_k)`, we set `result_k=result_k*p(p_k)`
+/// which fits better to how we use `p(p_k)`
+pub fn lagrange_evaluation<F>(x: &[F], y: &[F], p: &[F], result: &mut [F]) -> ()
```

Comment:
I think this is something more than just evaluating a polynomial at given points. It is probably tied to what malicious security needs to do, so it could be broken up into two parts - polynomial evaluation and everything else